test/acceptance/framework/consul_cluster.go

```diff
@@ -0,0 +1,151 @@
+package framework
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/gruntwork-io/terratest/modules/helm"
+	"github.com/gruntwork-io/terratest/modules/k8s"
+	"github.com/gruntwork-io/terratest/modules/logger"
+	"github.com/hashicorp/consul-helm/test/acceptance/helpers"
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/sdk/freeport"
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+// Cluster represents a consul cluster object
+type Cluster interface {
+	Create(t *testing.T)
+	Destroy(t *testing.T)
+	Upgrade(t *testing.T)
+	SetupConsulClient(t *testing.T, secure bool) *api.Client
+}
+
+// HelmCluster implements Cluster and uses Helm
+// to create, destroy, and upgrade consul
+type HelmCluster struct {
+	helmOptions *helm.Options
+	releaseName string
+	kubernetesClient *kubernetes.Clientset
+}
+
+func NewHelmCluster(
+	t *testing.T,
+	helmValues map[string]string,
+	ctx TestContext,
+	cfg *TestConfig,
+	releaseName string,) Cluster {
+
+	// Deploy single-server cluster by default unless helmValues overwrites that
+	values := map[string]string{
+		"server.replicas": "1",
+		"server.bootstrapExpect": "1",
+	}
+	valuesFromConfig := cfg.HelmValuesFromConfig()
+
+	// Merge all helm values
+	mergeMaps(values, valuesFromConfig)
+	mergeMaps(values, helmValues)
+
+	opts := &helm.Options{
+		SetValues:      values,
+		KubectlOptions: ctx.KubectlOptions(),
+		Logger:         logger.TestingT,
+	}
+	return &HelmCluster{
+		helmOptions: opts,
+		releaseName: releaseName,
+		kubernetesClient: ctx.KubernetesClient(t),
+	}
+}
+
+func (h *HelmCluster) Create(t *testing.T) {
```

Comment:
Can this cleanup old envs or error out if there's already something running? I keep running into an issue where I Ctrl-C the tests, it doesn't clean up properly and then when I re-run the test it fails again because there's already a server running and it's taking up the only port.

Reply:
I like this idea! I think erroring out is probably the best so that we don't accidentally delete anything you might actually need.

Reply:
I've added erroring and also set up things so that they are cleaned up when you ctrl-c. I've added `helpers.Cleanup` function that can be called instead of `t.Cleanup`. This function will set up an interrupt handler that will call the same cleanup function in case an interrupt signal is caught. Let me know what you think!